### PR TITLE
refactor(DbSlice): Replace `FindExt()` with `FindMutable()` and `FindReadOnly`

### PR DESCRIPTION
**Pull request by chakaz**

### CODE REVIEW

File: src/server/db_slice.h [hunk 1]
```diff
@@ -190,18 +190,21 @@ class DbSlice {
...
+    ExpireIterator exp_it;
     AutoUpdater post_updater;
   };
+  ItAndUpdater FindMutable(const Context& cntx, std::string_view key);
   OpResult<ItAndUpdater> FindMutable(const Context& cntx, std::string_view key,
                                      unsigned req_obj_type);
 
+  struct ItAndExpConst {
+    PrimeConstIterator it;
+    ExpireIterator exp_it;
```
Comment (adiholden) [inline src/server/db_slice.h:202]: ExpireConstIterator ?

Comment (chakaz) [reply src/server/db_slice.h]: Good idea, done :)

File: src/server/db_slice.cc [hunk 2]
```diff
@@ -400,15 +400,28 @@ DbSlice::AutoUpdater::AutoUpdater(const Fields& fields) : fields_(fields) {
...
 DbSlice::AddOrFindResult& DbSlice::AddOrFindResult::operator=(ItAndUpdater&& o) {
   it = o.it;
-  exp_it = ExpireIterator{};  // ItAndUpdater doesn't have exp_it
+  exp_it = o.exp_it;
   is_new = false;
   post_updater = std::move(o).post_updater;
   return *this;
 }
 
+DbSlice::ItAndUpdater DbSlice::FindMutable(const Context& cntx, string_view key) {
```
Comment (adiholden) [inline src/server/db_slice.cc:409]: I suggest instead of duplicating the code with the other find mutable that checks the type you can use OBJ_TYPE_MAX to indicate inside find mutable that you will not check the type, and you can have this set by default so the calls to this function which dont care about obj type will send only cntx and key

File: src/server/db_slice.cc [hunk 3]
```diff
@@ -400,15 +400,28 @@ DbSlice::AutoUpdater::AutoUpdater(const Fields& fields) : fields_(fields) {
...
+  } else {
+    return {it, exp_it, {}};
+  }
+}
+
 OpResult<DbSlice::ItAndUpdater> DbSlice::FindMutable(const Context& cntx, string_view key,
                                                      unsigned req_obj_type) {
-  auto it = FindInternal(cntx, key, FindInternalMode::kDontUpdateCacheStats).first;
+  // Don't use FindMutable() so that we don't call PreUpdate()
+  auto [it, exp_it] = FindInternal(cntx, key, FindInternalMode::kUpdateCacheStats);
```
Comment (adiholden) [inline src/server/db_slice.cc:424]: why did you change to kUpdateCacheStats?

File: src/server/db_slice.cc (hunk 2 above)
Comment (chakaz) [reply src/server/db_slice.cc:409]: This will be a big change, as the `type`-accepting versions return an `optional`, so it means that every call to previous `FindExt()` will now have to check `.value()`, which is not always desired..

File: src/server/db_slice.cc [hunk 4]
```diff
@@ -418,42 +431,50 @@ OpResult<DbSlice::ItAndUpdater> DbSlice::FindMutable(const Context& cntx, string
...
-      {it, AutoUpdater({AutoUpdater::DestructorAction::kRun, this, cntx.db_index, it, key, true})}};
+  return {{it, exp_it,
+           AutoUpdater({AutoUpdater::DestructorAction::kRun, this, cntx.db_index, it, key, true})}};
 }
 
-auto DbSlice::FindReadOnly(const Context& cntx, string_view key, unsigned req_obj_type) const
-    -> OpResult<PrimeConstIterator> {
-  auto res = Find(cntx, key, req_obj_type);
-  return res.ok() ? OpResult<PrimeConstIterator>(res.value()) : res.status();
+DbSlice::ItAndExpConst DbSlice::FindReadOnly(const Context& cntx, std::string_view key) const {
```
Comment (adiholden) [inline src/server/db_slice.cc:438]: same comment on object type here

File: src/server/db_slice.cc (hunk 3 above)
Comment (chakaz) [reply src/server/db_slice.cc:424]: Sorry, my bad. It was before I moved the implementation of read-only `GET` to use `FindReadOnly()`. Reverted.

File: src/server/db_slice.cc [hunk 5]
```diff
@@ -330,9 +330,9 @@ void DbSlice::Reserve(DbIndex db_ind, size_t key_size) {
   db->prime.Reserve(key_size);
 }
 
-auto DbSlice::Find(const Context& cntx, string_view key, unsigned req_obj_type) const
-    -> OpResult<PrimeIterator> {
-  auto it = FindInternal(cntx, key, FindInternalMode::kDontUpdateCacheStats).first;
+OpResult<PrimeIterator> DbSlice::Find(const Context& cntx, string_view key,
```
Comment (adiholden) [inline src/server/db_slice.cc:333]: why we dont have pre update in this flow?

Comment (chakaz) [reply src/server/db_slice.cc:333]: `Find()` is the last function to be converted. Once we merge this PR, the only API that will be using it is `FindFirst()`, which I will work on in a next PR.

File: src/server/generic_family.cc [hunk 6]
```diff
@@ -351,8 +351,8 @@ void Renamer::Finalize(Transaction* t, bool skip_exist_dest) {
 
 OpStatus Renamer::MoveSrc(Transaction* t, EngineShard* es) {
   if (es->shard_id() == src_sid_) {  // Handle source key.
-    // TODO: to call PreUpdate/PostUpdate.
-    auto it = es->db_slice().FindExt(t->GetDbContext(), src_res_.key).first;
+    auto res = es->db_slice().FindMutable(t->GetDbContext(), src_res_.key);
+    auto it = res.it;
```
Comment (adiholden) [inline src/server/generic_family.cc:355]: auto&

File: src/server/generic_family.cc [hunk 7]
```diff
@@ -379,8 +380,8 @@ OpStatus Renamer::UpdateDest(Transaction* t, EngineShard* es) {
   if (es->shard_id() != src_sid_) {
     auto& db_slice = es->db_slice();
     string_view dest_key = dest_res_.key;
-    PrimeIterator dest_it = db_slice.FindExt(t->GetDbContext(), dest_key).first;
-    DbSlice::AutoUpdater post_updater;
+    auto res = db_slice.FindMutable(t->GetDbContext(), dest_key);
+    auto dest_it = res.it;
```
Comment (adiholden) [inline src/server/generic_family.cc:384]: auto&

File: src/server/generic_family.cc (hunk 6 above)
Comment (chakaz) [reply src/server/generic_family.cc:355]: Copying iterators is trivial and we do it all the time (it has 2 pointers and 2 bytes), is there any particular reason to use a ref here?

Comment (chakaz) [reply src/server/generic_family.cc]: Anyway, done :)

File: src/server/generic_family.cc (hunk 7 above)
Comment (chakaz) [reply src/server/generic_family.cc]: Done.

File: src/server/generic_family.cc [hunk 8]
```diff
@@ -398,10 +399,7 @@ OpStatus Renamer::UpdateDest(Transaction* t, EngineShard* es) {
       if (src_res_.ref_val.ObjType() == OBJ_STRING) {
         pv_.SetString(str_val_);
       }
-      auto add_res =
-          db_slice.AddNew(t->GetDbContext(), dest_key, std::move(pv_), src_res_.expire_ts);
-      dest_it = add_res.it;
```
Comment (adiholden) [inline src/server/generic_family.cc:403]: I belive you should update dest_it

Comment (chakaz) [reply src/server/generic_family.cc:403]: but `dest_it` is now a reference to `res.it`

Comment (chakaz) [reply src/server/generic_family.cc:403]: (also - we can't update references to point to another object)

File: src/server/generic_family.cc [hunk 9]
```diff
@@ -515,7 +516,7 @@ OpResult<bool> OnRestore(const OpArgs& op_args, std::string_view key, std::strin
 bool ScanCb(const OpArgs& op_args, PrimeIterator it, const ScanOpts& opts, StringVec* res) {
   auto& db_slice = op_args.shard->db_slice();
   if (it->second.HasExpire()) {
-    it = db_slice.ExpireIfNeeded(op_args.db_cntx, it).first;
```
Comment (adiholden) [inline src/server/generic_family.cc:518]: I think we are missing a preupdate call on the scan items. If we expire the entry while scanning than how do we make sure we call PreUpdate before we delete it?

File: src/server/generic_family.cc [hunk 10]
```diff
@@ -1437,54 +1440,54 @@ OpResult<void> GenericFamily::OpRen(const OpArgs& op_args, string_view from_key,
...
+  if (IsValid(to_res.it)) {
+    to_res.it->second = std::move(from_obj);
+    to_res.it->second.SetExpire(IsValid(to_res.exp_it));  // keep the expire flag on 'to'.
 
     // It is guaranteed that UpdateExpire() call does not erase the element because then
     // from_it would be invalid. Therefore, UpdateExpire does not invalidate any iterators,
     // therefore we can delete 'from_it'.
-    db_slice.UpdateExpire(op_args.db_cntx.db_index, to_it, exp_ts);
-    CHECK(db_slice.Del(op_args.db_cntx.db_index, from_it));
+    db_slice.UpdateExpire(op_args.db_cntx.db_index, to_res.it, exp_ts);
```
Comment (adiholden) [inline src/server/generic_family.cc:1476]: In OpExpire you call post_updater.Run() berfor you call UpdateExpire. So is it needed to run post udpated before calling UpdateExpire? If so we should add it here as well

File: src/server/generic_family.cc [hunk 11]
```diff
@@ -1497,9 +1500,9 @@ OpResult<uint32_t> GenericFamily::OpStick(const OpArgs& op_args, ArgSlice keys)
 
   uint32_t res = 0;
   for (uint32_t i = 0; i < keys.size(); ++i) {
-    auto [it, _] = db_slice.FindExt(op_args.db_cntx, keys[i]);
-    if (IsValid(it) && !it->first.IsSticky()) {
-      it->first.SetSticky(true);
+    auto find_res = db_slice.FindMutable(op_args.db_cntx, keys[i]);
```
Comment (adiholden) [inline src/server/generic_family.cc:1503]: I suspect we have a bug here that we dont remove the expire info if there is

File: src/server/generic_family.cc [hunk 12]
```diff
@@ -1514,28 +1517,29 @@ OpStatus GenericFamily::OpMove(const OpArgs& op_args, string_view key, DbIndex t
...
+  auto from_res = db_slice.FindMutable(op_args.db_cntx, key);
+  if (!IsValid(from_res.it))
     return OpStatus::KEY_NOTFOUND;
 
   // Fetch value at key in target db.
   DbContext target_cntx = op_args.db_cntx;
   target_cntx.db_index = target_db;
-  auto [to_it, _] = db_slice.FindExt(target_cntx, key);
-  if (IsValid(to_it))
+  auto to_res = db_slice.FindMutable(target_cntx, key);
```
Comment (adiholden) [inline src/server/generic_family.cc:1527]: I think this should be ReadOnly we just check the to_res is valid

File: src/server/generic_family.cc (hunk 11 above)
Comment (chakaz) [reply src/server/generic_family.cc:1503]: 1. You mean from the expiry dash table, right?
2. Good catch!
3. That's regardless of this PR, right?

File: src/server/generic_family.cc (hunk 10 above)
Comment (chakaz) [reply src/server/generic_family.cc:1476]: `UpdateExpire()` can call `Del()` inside, and that's why we need to invoke the post updater before.
But see the comment in the above line: "It is guaranteed that UpdateExpire() call does not erase the element because..."
(if it would delete, then the call to `Del()` below is bad, I guess?)

File: src/server/generic_family.cc (hunk 12 above)
Comment (chakaz) [reply src/server/generic_family.cc]: Nice catch, done.

File: src/server/generic_family.cc [hunk 13]
```diff
@@ -1514,28 +1517,29 @@ OpStatus GenericFamily::OpMove(const OpArgs& op_args, string_view key, DbIndex t
...
+  bool sticky = from_res.it->first.IsSticky();
+  uint64_t exp_ts = db_slice.ExpireTime(from_res.exp_it);
+  PrimeValue from_obj = std::move(from_res.it->second);
 
   // Restore expire flag after std::move.
-  from_it->second.SetExpire(IsValid(from_expire));
+  from_res.it->second.SetExpire(IsValid(from_res.exp_it));
 
-  CHECK(db_slice.Del(op_args.db_cntx.db_index, from_it));
+  from_res.post_updater.Run();
```
Comment (adiholden) [inline src/server/generic_family.cc:1541]: In line 1536 we call std::move on the value. I feel we have a bug here as we restore only the expire flag after that. What if we have other flags turned on such as external entry. What will update in post update run? will it count the correct values?

File: src/server/json_family.cc [hunk 14]
```diff
@@ -591,7 +591,7 @@ OpResult<long> OpDel(const OpArgs& op_args, string_view key, string_view path) {
   long total_deletions = 0;
   if (path.empty()) {
     auto& db_slice = op_args.shard->db_slice();
-    auto [it, _] = db_slice.FindExt(op_args.db_cntx, key);
+    auto it = db_slice.FindMutable(op_args.db_cntx, key).it;  // We can ignore post_updater
```
Comment (adiholden) [inline src/server/json_family.cc:594]: why we can ignore post upater? what does it mean ignore?

File: src/server/set_family.cc [hunk 15]
```diff
@@ -559,7 +559,7 @@ OpResult<uint32_t> OpAdd(const OpArgs& op_args, std::string_view key, ArgSlice v
   // to overwrite the key. However, if the set is empty it means we should delete the
   // key if it exists.
   if (overwrite && vals.empty()) {
-    auto it = db_slice.FindExt(op_args.db_cntx, key).first;
+    auto it = db_slice.FindMutable(op_args.db_cntx, key).it;  // We can ignore post_updater
```
Comment (adiholden) [inline src/server/set_family.cc:562]: same quetion, why ignore?

File: src/server/string_family.cc [hunk 16]
```diff
@@ -183,26 +183,27 @@ OpResult<bool> ExtendOrSkip(const OpArgs& op_args, string_view key, string_view
...
 }
 
-OpResult<string> OpGet(const OpArgs& op_args, string_view key, bool del_hit = false,
-                       const DbSlice::ExpireParams& exp_params = {}) {
-  /*Get primeIterator and ExpireIterator at the same time*/
-  auto [it, it_expire] = op_args.shard->db_slice().FindExt(op_args.db_cntx, key);
+OpResult<string> OpMutableGet(const OpArgs& op_args, string_view key, bool del_hit = false,
+                              const DbSlice::ExpireParams& exp_params = {}) {
+  auto res = op_args.shard->db_slice().FindMutable(op_args.db_cntx, key);
+  res.post_updater.Run();
```
Comment (adiholden) [inline src/server/string_family.cc:189]: why do you call post updater here?

File: src/server/zset_family.cc [hunk 17]
```diff
@@ -962,7 +962,7 @@ OpResult<AddResult> OpAdd(const OpArgs& op_args, const ZParams& zparams, string_
   auto& db_slice = op_args.shard->db_slice();
 
   if (zparams.override && members.empty()) {
-    auto it = db_slice.FindExt(op_args.db_cntx, key).first;
+    auto it = db_slice.FindMutable(op_args.db_cntx, key).it;  // We can ignore post_updater
```
Comment (adiholden) [inline src/server/zset_family.cc:965]: also why here ignore?

File: src/server/generic_family.cc (hunk 8 above)
Comment (adiholden) [reply src/server/generic_family.cc:403]: ops you are right

File: src/server/generic_family.cc (hunk 10 above)
Comment (adiholden) [reply src/server/generic_family.cc:1476]: Very confusing...

File: src/server/generic_family.cc (hunk 11 above)
Comment (adiholden) [reply src/server/generic_family.cc:1503]: yes from the expire dash table. Sure this bug was not introduced by this PR

File: src/server/generic_family.cc (hunk 9 above)
Comment (adiholden) [reply src/server/generic_family.cc:518]: I believe we have the same problem with Eviction and Expiring items on background process

File: src/server/generic_family.cc (hunk 13 above)
Comment (chakaz) [reply src/server/generic_family.cc:1541]: This is an existing bug, right?
I'm not sure how `PostUpdate()` will handle it, but it might be a good idea to load from memory both src and dest. Is this not part of the flow anyway?

File: src/server/json_family.cc (hunk 14 above)
Comment (chakaz) [reply src/server/json_family.cc:594]: By ignore I meant that we can let `PostUpdate()` run immediately. I'm open to rephrasing however you think will be clear.

File: src/server/set_family.cc (hunk 15 above)
Comment (chakaz) [reply src/server/set_family.cc:562]: See above.

File: src/server/generic_family.cc (hunk 10 above)
Comment (chakaz) [reply src/server/generic_family.cc:1476]: Shall I move the `Run()` call to be before `UpdateExpire()`? I guess there's no harm in that, will it be less confusing?

File: src/server/generic_family.cc (hunk 11 above)
Comment (chakaz) [reply src/server/generic_family.cc:1503]: Filed https://github.com/dragonflydb/dragonfly/issues/2315

File: src/server/string_family.cc (hunk 16 above)
Comment (chakaz) [reply src/server/string_family.cc:189]: We don't update the value at any point - we either (maybe) delete it, or (maybe) expire it. Either way, no harm in calling it now.
However, now that you mention it, the call below (line 205) is redundant. I'll remove it.

File: src/server/generic_family.cc (hunk 13 above)
Comment (adiholden) [reply src/server/generic_family.cc:1541]: This is an existing bug.
1. We can run the post update on from before we are doting the std::move we are not going to update it but delete it next
2. I am not sure how the db_slice.Del works after we call the std::move , we can create a github issue for this and add there to add a check that we are not calling move on io pending /external entry as this is not supported in this flow.

File: src/server/generic_family.cc [hunk 18]
```diff
@@ -1437,54 +1440,54 @@ OpResult<void> GenericFamily::OpRen(const OpArgs& op_args, string_view from_key,
...
-  if (!IsValid(from_it))
+  auto from_res = db_slice.FindMutable(op_args.db_cntx, from_key);
+  if (!IsValid(from_res.it))
     return OpStatus::KEY_NOTFOUND;
 
   if (from_key == to_key)
     return OpStatus::OK;
 
   bool is_prior_list = false;
   DbSlice::AutoUpdater post_updater;
```
Comment (adiholden) [inline src/server/generic_family.cc:1451]: I guess you can remove this post_updater

File: src/server/generic_family.cc (hunk 10 above)
Comment (adiholden) [reply src/server/generic_family.cc:1476]: Oh I see here to we call std::move on the value, again here I dont know how this is working

File: src/server/generic_family.cc (hunk 9 above)
Comment (adiholden) [reply src/server/generic_family.cc:518]: lets open another github issue on this

File: src/server/generic_family.cc (hunk 13 above)
Comment (chakaz) [reply src/server/generic_family.cc:1541]: Moved to before the `move()`.
What's wrong with `Del()` after `move()`?

File: src/server/generic_family.cc (hunk 9 above)
Comment (chakaz) [reply src/server/generic_family.cc:518]: I remember writing a reply but I don't see it now.
I think this should be handled in ExpireIfNeeded(). I'll file a bug.

Comment (chakaz) [reply src/server/generic_family.cc:518]: https://github.com/dragonflydb/dragonfly/issues/2316

File: src/server/generic_family.cc (hunk 10 above)
Comment (chakaz) [reply src/server/generic_family.cc:1475]: What specifically is unclear?

**Review by adiholden — APPROVED** [review]
